### PR TITLE
Disable sharp corner case when no dasharray/gradient/pattern is active

### PR DESCRIPTION
**Pull request by rreusser**
This PR attempts a partial remedy for https://github.com/mapbox/mapbox-gl-js/issues/10000. The problem is that extra vertices are inserted at sharp corners (> 75 degrees) so that dashes, gradients, and patterns don't get skewed by corner miters.

However, this vertex insertion causes self-intersecting geometry at corners, which prevents adding transparency to lines (via color or via line-blur) without artifacts, as seen in #10000. This PR aims for a reasonably solution by which we simply disable this vertex insertion when there's no pattern and thus nothing to be gained.

An additional downside of this insertion at layout time is that since the width of the line is not known until paint time, its implementation introduces a scale-dependence and thus disproportionately affects wider lines, making the line-blur property particularly difficult to use, since blurred lines tend to be somewhat wider.

The code of interest is:

https://github.com/mapbox/mapbox-gl-js/blob/7d2f272e4cf478c286fb115cc7550a19ea477f91/src/data/bucket/line_bucket.js#L47-L59

The images below illustrate the fact that these artifacts depend on _line width_ and not only _line geometry_, even at an isolated corner.

Narrow lines:
Screen Shot 2021-08-11 at 8 40 45 AM

Same setup, just a wider line:
Screen Shot 2021-08-11 at 8 40 30 AM

I've added test:
Previous behavior:

expected-orig

Updated behavior:

expected

Questions:

- I'm not sure whether I'm correctly detecting the presence of gradients or dasharray correctly.
- This might also require checking for a pattern. Are there other cases? Are there so many cases that this is not advisable at all?

cc @mapbox/gl-native. Whether this change is desirable remains to be seen, but if so, it would probably be appropriate to mirror in gl-native.

## Launch Checklist

 - [x] briefly describe the changes in this PR
 - [x] include before/after visuals or gifs if this PR includes visual changes
 - [x] write tests for all new functionality
 - [ ] document any changes to public APIs
 - [ ] post benchmark scores
 - [ ] manually test the debug page
 - [ ] tagged `@mapbox/map-design-team` `@mapbox/static-apis` if this PR includes style spec API or visual changes
 - [ ] tagged `@mapbox/gl-native` if this PR includes shader changes or needs a native port
 - [x] apply changelog label ('bug', 'feature', 'docs', etc) or use the label 'skip changelog'
 - [x] add an entry inside this element for inclusion in the `mapbox-gl-js` changelog: `<changelog>Improve rendering of partially transparent lines with large width</changelog>`

### CODE REVIEW

**Comment by rreusser** [comment]
One additional note: I also tried using the winding number (CW/CCW) to cull triangles which flip when they start to self-intersect. I thought it had a good chance of helping, but it turned out this particular behavior does not actually flip any triangles. I think it might help though in the case of two miters which intersect and _do_ flip the triangles. Though culling certainly can't simply be flipped on without a lot of thought.

File: src/data/bucket/line_bucket.js [hunk 1]
```diff
@@ -436,7 +444,7 @@ class LineBucket implements Bucket {
             // approximate angle from cosine
             const approxAngle = 2 * Math.sqrt(2 - 2 * cosHalfAngle);
 
-            const isSharpCorner = cosHalfAngle < COS_HALF_SHARP_CORNER && prevVertex && nextVertex;
+            const isSharpCorner = cosHalfAngle < COS_HALF_SHARP_CORNER && prevVertex && nextVertex && adjustSharpCorners;
```
Comment (endanke) [inline src/data/bucket/line_bucket.js:447]: nit: Might be good to rename `isSharpCorner` to something else, because the corner is still sharp even if `adjustSharpCorners` is false. Maybe: `needsSharpCornerAdjustment`

File: src/data/bucket/line_bucket.js [hunk 2]
```diff
@@ -319,16 +319,24 @@ class LineBucket implements Bucket {
         const cap = layout.get('line-cap').evaluate(feature, {});
         const miterLimit = layout.get('line-miter-limit');
         const roundLimit = layout.get('line-round-limit');
+        const dashArray = this.layers[0].paint.get('line-dasharray').value;
```
Comment (endanke) [inline src/data/bucket/line_bucket.js:322]: Don't we need to check all layers to see if at least one of them has these properties? In gl-native I wrote a check for gradients which loops through the layers in the constructor of the bucket and sets a flag if any of them has it. 
Although I'm not fully sure if it's needed.

https://github.com/mapbox/mapbox-gl-native-internal/blob/internal/src/mbgl/renderer/buckets/line_bucket.cpp#L30

Comment (rreusser) [reply src/data/bucket/line_bucket.js:322]: Yes! I was confused by this line: `layout = this.layers[0].layout` Maybe all layers point to the same singleton layout?

https://github.com/mapbox/mapbox-gl-js/blob/f28733a812274dd25e838200b7bd7cf4bf20095f/src/data/bucket/line_bucket.js#L317

Exactly what needs to be checked was my main question in this PR. I'll take a look at your code and work through this again to see what needs to be checked here.

Comment (alexshalamov) [reply src/data/bucket/line_bucket.js:322]: @rreusser code has changed a bit, but afaik, layers are grouped by layout https://github.com/mapbox/mapbox-gl-js/blob/10961e05522d38bae952d7f01947fe6bfd6f2aa1/src/style/style_layer_index.js#L53.
@endanke is right, we might need to go through paint properties of all layers in a group.

File: test/integration/render-tests/line-dasharray/sharp-lines/style.json [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+      }
+    },
+    {
+      "id": "sharp-line",
+      "type": "line",
+      "source": "geojson",
+      "paint": {
+        "line-color": "rgba(0, 0, 255, 0.5)",
+        "line-width": 50
+      }
```
Comment (alexshalamov) [inline test/integration/render-tests/line-dasharray/sharp-lines/style.json:45]: Would be nice to add one more layer with pattern / gradient / dash array, to test https://github.com/mapbox/mapbox-gl-js/pull/10925/files#r701327734 case. When layer with the same layout has properties that can affect calculation of `adjustSharpCorners` variable.

File: src/data/bucket/line_bucket.js [hunk 4]
```diff
@@ -319,16 +319,24 @@ class LineBucket implements Bucket {
...
+        const hasDashArray = dashArray.kind !== 'constant' || !!dashArray.value;
+        const hasGradient = !!this.layers[0].paint.get('line-gradient');
+        const hasPattern = pattern.kind !== 'constant' || !!pattern.value;
+        const adjustSharpCorners = hasDashArray || hasGradient || hasPattern;
+
         this.lineClips = this.lineFeatureClips(feature);
 
         for (const line of geometry) {
-            this.addLine(line, feature, join, cap, miterLimit, roundLimit);
+            this.addLine(line, feature, join, cap, miterLimit, roundLimit, adjustSharpCorners);
```
Comment (alexshalamov) [inline src/data/bucket/line_bucket.js:333]: 🤔 I think there is an architectural problem with the approach. In most of the cases, bucket's data that is created at layout phase, can't depend on paint properties. Can be tested by creating line layer and then setting dash array or gradient paint property using runtime apis.

Comment (rreusser) [reply src/data/bucket/line_bucket.js:333]: Thanks for pointing this out, @alexshalamov! That makes perfect sense, though I do observe that this layout code is re-executed when I change the `line-dasharray`, e.g.:

```
map.setPaintProperty('path', 'line-dasharray', [2, 1])
```

Comment (rreusser) [reply src/data/bucket/line_bucket.js:333]: Ah, now I see. I think my success was dumb luck, as I only tested this situation with `line-dasharray`. My current hypothesis is that the data-driven nature of `line-dasharray` causes re-execution of the line bucket layout code, which was sufficient to trigger this to update. As `line-gradient` is *not* data-driven, the gradient paint property does not trigger the update and the state is not correctly updated.

In summary, I think you're right that this is a fundamental error in my approach. I think it's necessary to simply close this PR since it's just too ugly of a hack which, more importantly, doesn't cover all cases.

**Comment by rreusser** [comment]
@alexshalamov points out that the properties which affect this toggle are *paint* properties, thus changing them at runtime won't trigger re-*layout*, and the geometry won't be correctly updated. It's my experience that it *does*, in fact, work with `line-dasharray`, but I believe it's merely dumb luck since it's the fact that `line-dasharray` is data-driven in GL JS (I _think_; this is my hypothesis) that triggers the update. I believe this won't work in native since it's not data-driven, and I've definitely confirmed that the analogous fix does not correctly update for `line-gradient`, which is not data-driven.

I think it's necessary to simply close this PR as a reasonable attempt that's ultimately fundamentally incompatible with the layout -> paint -> render pipeline.

cc @dudeuter
